File: number_theory/03_prime_factorization.py


def prime_factors(n : int) -> list:
	factors = []
	d = 2;
	while( d*d <= n ):
		while( n % d == 0 ):
			factors.append(d)
			n //= d
		d += 1
	if n > 1:
		factors.append(n)
	return factors


n = int(input("Enter number : "))
factors = prime_factors(n)
if( not factors ):
	print("No prime factors.")
for val in factors:
	print(val, end=' ')
print()

File: number_theory/03_prime_factorization.cpp
#include<iostream>
#include<vector>
using namespace std;

vector<int> prime_factors(int n){
	vector<int> factors;
	for( int d = 2; d*d <= n; d++ ){
		while( n % d == 0 ){
			factors.push_back(d);
			n /= d;
		}
	}
	if( n > 1 )
		factors.push_back(n);
	return factors;
}

int main(){

	int n;
	cout << "Enter number to find its prime factors : ";
	cin >> n;
	vector<int> factors = prime_factors(n);
	for( int val : factors )
		cout << val << " ";
	if( factors.size() == 0 )
		cout << "No prime factors.";
	cout << endl;

	return 0;
}